core/test/Cardano/ScriptSpec.hs

```diff
@@ -251,6 +261,34 @@ spec = do
                         ]
                     ]
             validateScript script `shouldBe` Right ()
+
+    describe "can perform roundtrip JSON serialization & deserialization" $
+        it "fromJSON . toJSON === pure" $ do
+            let roundtrip =
+                    Json.decode . Json.encode . toJSON :: Script -> Maybe Script
+            property $ \(t :: Script) -> roundtrip t === pure t
+
   where
     toHexText = T.decodeUtf8 . encode EBase16
     toHexText' (ScriptHash bytes) = toHexText bytes
+
+instance Arbitrary Script where
+    arbitrary = do
```

Comment:
I can't remember ... with generators of recursive types, does the generator need to be made smaller for every step down the tree?

Reply:
:thinking: